src/object/cli_obj.c

```diff
@@ -2444,19 +2437,419 @@ obj_req_fanout(struct dc_object *obj, struct obj_auxi_args *obj_auxi,
+	shard_arg->la_kds = NULL;
+	if (shard_arg->la_sgl != NULL && shard_arg->la_sgl != obj_arg->sgl)
+		daos_sgl_fini(shard_arg->la_sgl, false);
+
+	return 0;
+}
+
+static void
+obj_auxi_list_fini(struct obj_auxi_args *obj_auxi)
+{
+	tse_task_list_traverse(&obj_auxi->shard_task_head,
+			       shard_list_task_fini, obj_auxi);
+}
+
+struct comp_cb_merge_arg {
+	d_list_t	merge_list;
+	int		merge_nr;
+	daos_off_t	merge_sgl_off;
+};
+
+static int
+merge_recx_insert(d_list_t *prev, uint64_t offset, uint64_t size)
+{
+	struct obj_auxi_list_recx *new;
+
+	D_ALLOC_PTR(new);
+	if (new == NULL)
+		return -DER_NOMEM;
+
+	new->recx.rx_idx = offset;
+	new->recx.rx_nr = size;
+	D_INIT_LIST_HEAD(&new->recx_list);
+	d_list_add_tail(&new->recx_list, prev);
+	return 0;
+}
+
+int
+merge_recx(d_list_t *head, uint64_t offset, uint64_t size)
+{
+	struct obj_auxi_list_recx	*recx;
+	struct obj_auxi_list_recx	*tmp;
+	struct obj_auxi_list_recx	*prev = NULL;
+	bool				inserted = false;
+	int				rc = 0;
+
+	d_list_for_each_entry_safe(recx, tmp, head, recx_list) {
+		if (offset + size < recx->recx.rx_idx) {
+			rc = merge_recx_insert(prev == NULL ?
+					       head : &prev->recx_list,
+					       offset, size);
+			inserted = true;
+			break;
+		}
+
+		if (offset + size == recx->recx.rx_idx) {
+			recx->recx.rx_idx = offset;
+			recx->recx.rx_nr += size;
+			inserted = true;
+			break;
+		}
+
+		if (recx->recx.rx_idx + recx->recx.rx_nr == offset) {
+			recx->recx.rx_nr += size;
+			/* Try to merge with the next */
+			if (recx->recx_list.next != head) {
+				struct obj_auxi_list_recx *next;
+
+				next = d_list_entry(recx->recx_list.next,
+						    struct obj_auxi_list_recx,
+						    recx_list);
+				if (recx->recx.rx_idx + recx->recx.rx_nr ==
+				    next->recx.rx_idx) {
+					recx->recx.rx_nr += next->recx.rx_nr;
+					d_list_del(&next->recx_list);
+					D_FREE_PTR(next);
+				}
+			}
+			inserted = true;
+			break;
+		}
+		prev = recx;
+	}
+
+	if (!inserted)
+		rc = merge_recx_insert(prev == NULL ? head : &prev->recx_list,
+				       offset, size);
+
+	return rc;
+}
+
+static void
+obj_recx_parity_to_daos(struct daos_oclass_attr *oca, daos_recx_t *recx)
+{
+	uint64_t cur_off = recx->rx_idx & ~PARITY_INDICATOR;
 
+	recx->rx_idx = obj_ec_parity_off_to_daos_off(
+			cur_off, obj_ec_cell_rec_nr(oca),
+			obj_ec_stripe_rec_nr(oca));
+	recx->rx_nr = obj_ec_stripe_rec_nr(oca);
+}
+
+static int
+obj_ec_recxs_convert(d_list_t *merge_list, daos_recx_t *recx,
+		     struct shard_auxi_args *shard_auxi)
+{
+	struct daos_oclass_attr	*oca;
+	uint64_t		total_size = recx->rx_nr;
+	uint64_t		cur_off = recx->rx_idx & ~PARITY_INDICATOR;
+	int			rc = 0;
+
+	oca = daos_oclass_attr_find(shard_auxi->obj->cob_md.omd_id);
+	D_ASSERT(oca != NULL);
+	/* Normally the enumeration is sent to the parity node */
+	/* convert the parity off to daos off */
+	if (recx->rx_idx & PARITY_INDICATOR) {
+		obj_recx_parity_to_daos(oca, recx);
+		return 0;
+	}
+
+	if (merge_list == NULL)
+		return 0;
```

Comment:
in obj_enum_shard_process_cb(), it will call obj_ec_recxs_convert() with merge_list = NULL, which will be used in rebuild process, where we do not need merge the recx.